### PR TITLE
#517 Define routes for assessments feature 

### LINKED ISSUE #517: Expected format for each route of assessments feature (input + output) should be defined
### Describe the Feature / Enhancement

For each route defined in the list of routes that was in #506, the information being sent to the backend and retrieved from the backend needs to be defined in a structure so that we can wire up the backend to the frontend.

These data types will get added to the `api/src/models.d.ts` and `webapp/src/types/api.d.ts` files. However, the structure of these needs to be defined *first* before modifying these files.

### Additional Details and Resources

See here for reference of the routes: https://github.com/OpenTree-Education/rhizone-lms/pull/509

### Correct Project Selected

- [X] This issue is linked to the appropriate project

### Labels

- [X] This issue has appropriate labels.

### PR DESCRIPTION
**Pull request by Paryssatis**
## Proposed Changes

Resolves #517 For each route defined in the list of routes that was in #506, the information being sent to the backend and retrieved from the backend is now defined in a structure which allows us to wire up the backend to the frontend.

## Checklist

- [x] Related issue appears at beginning of pull request title with pound sign, and title describes the changes being proposed.
- [x] Related issue is linked in pull request description using [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] The appropriate label has been chosen for this pull request.
- [x] The correct project has been selected for this pull request.
- [x] All commits in this branch, including merge commits, begin with the issue number and a pound sign.
- [x] Tests have been added, where appropriate; or, no tests are relevant for this pull request.
- [x] This pull request contains UI changes, and screenshots of those UI images appear below; or, this pull request contains no UI changes.

### CODE REVIEW

File: api/src/models.d.ts [hunk 1]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
 }
+
+export interface Question {
+  assessment_question_id: number;
+  id?: number;
+  assessment_id?: number;
+  title: string;
+  description?: string;
+  question_type: string;
+  answers?: Answer[];
```
Comment (seidior) [inline api/src/models.d.ts:91]: `Answer` here is an object type, one that we also define in this file. As per our feedback earlier today, I would like you to have the interface definition for `Answer` appear above `Question` in this file; likewise, any other object types being referenced in any of the other interfaces should appear above them in this file.

File: api/src/models.d.ts [hunk 2]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
   program_id: number;
   participant_activities: ParticipantActivityCompletionStatus[];
 }
+
+export interface Question {
+  assessment_question_id: number;
+  id?: number;
```
Comment (seidior) [inline api/src/models.d.ts:86]: As mentioned previously, for all data types that have an `id` field, the `id` field should display first.

File: api/src/models.d.ts [hunk 3]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
   program_id: number;
   participant_activities: ParticipantActivityCompletionStatus[];
 }
+
+export interface Question {
+  assessment_question_id: number;
```
Comment (seidior) [inline api/src/models.d.ts]: This field does not appear in the data types Google Doc I'm referencing. What is this field, and what information does it reference?

File: api/src/models.d.ts [hunk 4]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  id?: number;
+  title: string;
+  description?: string;
+  max_score: number;
+  max_num_submissions: number;
+  time_limit?: number;
+  curriculum_id: number;
+  activity_id: number;
+  principal_id: number;
+  questions: Question[];
```
Comment (seidior) [inline api/src/models.d.ts]: This parameter should be optional.

File: api/src/models.d.ts [hunk 5]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  num_participants_with_submissions: number;
+  num_program_participants: number;
+  num_ungraded_submissions: number;
+}
+
+export interface ProgramAvailableAssessmentsSummary {
+  curriculum_assessment: CurriculumAssessment;
+  program_assessment: ProgramAssessment;
+  submissions_summary:
+    | AssessmentSubmissionsSummary
```
Comment (seidior) [inline api/src/models.d.ts:136]: This field should not be prepended with the `|` (pipe) symbol. It should only appear between `AssessmentSubmissionsSummary` and `FacilitatorAssessmentSubmissionsSummary`.

File: api/src/models.d.ts [hunk 6]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  due_date: string;
+}
+
+export interface FacilitatorAssessmentSubmissionsSummary {
+  num_participants_with_submissions: number;
+  num_program_participants: number;
+  num_ungraded_submissions: number;
+}
+
+export interface ProgramAvailableAssessmentsSummary {
```
Comment (seidior) [inline api/src/models.d.ts]: Let's rename this `AssessmentSummary`.

File: api/src/models.d.ts [hunk 7]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  curriculum_assessment: CurriculumAssessment;
+  program_assessment: ProgramAssessment;
+  submissions_summary:
+    | AssessmentSubmissionsSummary
+    | FacilitatorAssessmentSubmissionsSummary;
+}
+
+export interface AssessmentResponses {
```
Comment (seidior) [inline api/src/models.d.ts]: Let's switch from the plural `AssessmentResponses` to the singular `AssessmentResponse`.

File: api/src/models.d.ts [hunk 8]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+
+export interface AssessmentSubmission {
+  id?: number;
+  assessment_id: number;
+  principal_id: number;
+  assessment_submission_state: string;
+  score?: number;
+  opened_at: string;
+  submitted_at?: string;
+  state_id: number;
```
Comment (seidior) [inline api/src/models.d.ts]: We already have `assessment_submission_state`; therefore, we don't need `state_id` as well.

File: api/src/models.d.ts [hunk 9]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  responses?: AssessmentResponses[];
+}
+
+export interface ProgramSubmittedAssessments {
```
Comment (seidior) [inline api/src/models.d.ts]: Let's rename this `SubmittedAssessment`.

File: api/src/models.d.ts [hunk 10]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  curriculum_assessment: CurriculumAssessment;
+  program_assessment: ProgramAssessment;
+  submissions: AssessmentSubmission[];
+}
+
+export interface AssessmentSubmission {
```
Comment (seidior) [inline api/src/models.d.ts]: We're already defining this on line 152 above. Is there a reason we need to define it again?

File: api/src/models.d.ts [hunk 11]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  assessment_id: number;
+  principal_id: number;
+  assessment_submission_state: string;
+  score?: number;
+  opened_at: string;
+  submitted_at?: string;
+  responses?: AssessmentResponses[];
+}
+
+export interface ProgramSubmittedAssessments {
```
Comment (seidior) [inline api/src/models.d.ts]: Let's remove this as well, since we're already defining it above on line 164.

File: api/src/models.d.ts [hunk 12]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  curriculum_assessment: CurriculumAssessment;
+  program_assessment: ProgramAssessment;
+  submission: AssessmentSubmission;
+}
+export interface AssessmentSubmission {
```
Comment (seidior) [inline api/src/models.d.ts]: Let's remove this, the third definition of AssessmentSubmission.

File: api/src/models.d.ts [hunk 13]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  assessment_id: number;
+  principal_id: number;
+  assessment_submission_state: string;
+  score?: number;
+  opened_at: string;
+  submitted_at?: string;
+  responses?: AssessmentResponses[];
+}
+
+export interface DraftProgramAssessment {
```
Comment (seidior) [inline api/src/models.d.ts]: Let's remove this as well. This is fundamentally the same data type as a `SubmittedAssessment` and therefore doesn't need another name.

File: api/src/models.d.ts [hunk 14]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  submission: AssessmentSubmission;
+}
+
+export interface ProgramParticipantCompletionSummary {
+  program: Program;
+  principal_id: number;
+  total_score: number;
+}
+
+export interface ProgramCertificate {
```
Comment (seidior) [inline api/src/models.d.ts]: We don't need a data type for this. This may be safely removed.

File: api/src/models.d.ts [hunk 15]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  completion_summary: ProgramParticipantCompletionSummary;
+}
+
+export interface NewProgramAssessment {
```
Comment (seidior) [inline api/src/models.d.ts]: We don't need a data type for this. This may be safely removed.

File: api/src/models.d.ts [hunk 16]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  curriculum_assessment: CurriculumAssessment;
+  program_assessment: ProgramAssessment;
+}
+
+export interface EditProgramAssessment {
```
Comment (seidior) [inline api/src/models.d.ts]: We don't need a data type for this. This may be safely removed.

File: api/src/models.d.ts [hunk 17]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  curriculum_assessment: CurriculumAssessment;
+  program_assessment: ProgramAssessment;
+}
+
+export interface DeleteProgramAssessment {
```
Comment (seidior) [inline api/src/models.d.ts]: We don't need a data type for this. This may be safely removed.

File: api/src/models.d.ts [hunk 18]
```diff
@@ -72,3 +72,158 @@ export interface ParticipantActivityForProgram {
...
+  submission: AssessmentSubmission;
+}
+
+export interface SubmitProgramAssessment {
```
Comment (seidior) [inline api/src/models.d.ts]: We don't need a data type for this. This may be safely removed.

**Review by seidior — CHANGES_REQUESTED** [review]
I'm only going to comment on `models.d.ts`, since your work is duplicated between the two files. For the most part, you've got the data types in place, so there's just a few minor comments to bring it closer to the spec.

You've also defined a few extraneous data types that were not called for in the document. If you notice, we don't define single-member data types anywhere else in the model file; when crafting code that is like other code that we have, make sure you follow the same design patterns and code style that we use elsewhere in our application.

Once you make the changes requested and you copy them over to the `api.d.ts` file, please let me know and I'll take another look at the file.

Thank you for your work on this!

**Review by seidior — APPROVED** [review]
You have adequately addressed the changes I have previously requested, and your code sufficiently addresses and resolves the issue in the ticket. I am approving it for merger with the feature branch.

Thanks so much team! Great work.

**Review by Awatanka — APPROVED** [review]
Great job with the models that were created for our router!

**Pull request merged**